File: ConcurrentJavaRollerCoaster/src/rafael/concurrent/monitor/rollercoaster/RollerCoasterOneCar.java
package rafael.concurrent.monitor.rollercoaster;

import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.ReentrantLock;

/**
 * This class is an implementation of The Roller Coaster concurrent problem for one car.
 * Invariant: (pc <= cc and cr => (pc == cc))
 *  - pc: number of passengers in the car
 *  - cc: number of passengers the car can carry
 *  - cr: true if the car is riding, false otherwise
 * @author Rafael de Castro Dantas Sales - rafaelcds@gmail.com
 */
public class RollerCoasterOneCar {
	
	protected final int CAR_CAPACITY;
	
	private int passengersInCar = 0; //Number of passengers in the car 
	private boolean carIsWaitingPassengers = false; //True if the car is waiting for passengers
	
	private final ReentrantLock rollerCoasterLock;
	private final Condition carLoad; //Signaled when the car can load passengers
	private final Condition carUnload; //Signaled when the car is unloading its passengers
	private final Condition carIsFull; //Signaled when the car is full
	
	/**
	 * Creates a Roller Coaster implementation using monitor.
	 * @param carCapacity number of passengers per car
	 */
	public RollerCoasterOneCar(int carCapacity) {
		this.CAR_CAPACITY = carCapacity;
		
		rollerCoasterLock = new ReentrantLock();
		carLoad = rollerCoasterLock.newCondition();
		carUnload = rollerCoasterLock.newCondition();
		carIsFull = rollerCoasterLock.newCondition();
	}
	
	/**
	 * Requests to take ride in the car.
	 * @param passengerId identifier of the passenger
	 */
	public void takeRide(final int passengerId) {
		rollerCoasterLock.lock();
		
		//Waits the car become available:
		while (!carIsWaitingPassengers || passengersInCar == CAR_CAPACITY) {
			System.out.println("Passenger " + passengerId + ": Waiting for the car");
			carLoad.awaitUninterruptibly();
		}
		
		passengersInCar++;
		System.out.println("Passenger " + passengerId + ": Will take ride in the car");
		if (passengersInCar == CAR_CAPACITY) {
			//Notifies the car that it is full:
			carIsFull.signal();
		}		
		//Waits until the car gets full:
		carIsFull.awaitUninterruptibly();
		//Waits the end of the ride:
		carUnload.awaitUninterruptibly();
		System.out.println("Passenger " + passengerId + ": Left the car");
		
		rollerCoasterLock.unlock();
	}
	
	/**
	 * Requests the car to carry passengers.
	 */
	public void load() {
		rollerCoasterLock.lock();
		
		//Waits for the car if it is riding:
		while (passengersInCar > 0) {
			carUnload.awaitUninterruptibly();
		}
		carIsWaitingPassengers = true;
		//Notifies the passengers the car is ready to load them:
		carLoad.signalAll();
		System.out.println("Car: Waiting for passengers");
		//Waits until the car gets full:
		carIsFull.awaitUninterruptibly();
		carIsWaitingPassengers = false;
		//Notifies the passengers that the car is full:
		carIsFull.signalAll();
		System.out.println("Car: The car is full. Going around...\n");
		
		rollerCoasterLock.unlock();
	}
	
	/**
	 * Unload the passengers from the car.
	 */
	public void unload() {
		rollerCoasterLock.lock();
		
		//Unload the passengers:
		carUnload.signalAll();
		passengersInCar = 0;
		System.out.println("Car: Unloaded its passengers");
		
		rollerCoasterLock.unlock();
	}
}

File: ConcurrentCppRollerCoaster/src/RollerCoasterOneCar.cpp
/*
 * RollerCoasterOneCar.cpp
 *
 *  Created on: May 9, 2010
 *      Author: Rafael de Castro Dantas Sales - rafaelcds@gmail.com
 */

#include <pthread.h>
#include <iostream>
#include "RollerCoasterOneCar.h"

using namespace std;

RollerCoasterOneCar::RollerCoasterOneCar(int carCapacity) : CAR_CAPACITY(carCapacity) {
	passengersInCar = 0;
	carIsWaitingPassengers = false;
	pthread_mutex_init(&rollerCoasterLock, NULL);
	pthread_cond_init(&carLoad, NULL);
	pthread_cond_init(&carUnload, NULL);
	pthread_cond_init(&carIsFull, NULL);
}

RollerCoasterOneCar::~RollerCoasterOneCar() {
}

void RollerCoasterOneCar::takeRide(const int passengerId) {
	pthread_mutex_lock(&rollerCoasterLock);

	//Waits the car become available:
	while (!carIsWaitingPassengers || passengersInCar == CAR_CAPACITY) {
		cout << "Passenger " << passengerId << ": Waiting for the car" << endl;
		pthread_cond_wait(&carLoad, &rollerCoasterLock);
	}

	passengersInCar++;
	cout << "Passenger " << passengerId << ": Will take ride in the car" << endl;
	if (passengersInCar == CAR_CAPACITY) {
		//Notifies the car that it is full:
		pthread_cond_signal(&carIsFull);
	}
	//Waits until the car gets full:
	pthread_cond_wait(&carIsFull, &rollerCoasterLock);
	//Waits the end of the ride:
	pthread_cond_wait(&carUnload, &rollerCoasterLock);
	cout << "Passenger " << passengerId << ": Left the car" << endl;

	pthread_mutex_unlock(&rollerCoasterLock);
}

void RollerCoasterOneCar::load() {
	pthread_mutex_lock(&rollerCoasterLock);

	//Waits for the car if it is riding:
	while (passengersInCar > 0) {
		pthread_cond_wait(&carUnload, &rollerCoasterLock);
	}
	carIsWaitingPassengers = true;
	//Notifies the passengers the car is ready to load them:
	pthread_cond_broadcast(&carLoad);
	cout << "Car: Waiting for passengers" << endl;
	//Waits until the car gets full:
	pthread_cond_wait(&carIsFull, &rollerCoasterLock);
	carIsWaitingPassengers = false;
	//Notifies the passengers that the car is full:
	pthread_cond_broadcast(&carIsFull);
	cout << "Car: The car is full. Going around...\n" << endl;

	pthread_mutex_unlock(&rollerCoasterLock);
}

void RollerCoasterOneCar::unload() {
	pthread_mutex_lock(&rollerCoasterLock);

	//Unload the passengers:
	pthread_cond_broadcast(&carUnload);
	passengersInCar = 0;
	cout << "Car: Unloaded its passengers" << endl;

	pthread_mutex_unlock(&rollerCoasterLock);
}
